proteus/NonlinearSolvers.py

```diff
@@ -755,26 +755,24 @@ def solve(self,u,r=None,b=None,par_u=None,par_r=None):
         self.F.secondCallCalculateResidual = 0
         self.computeResidual(u,r,b)
         u[:] = r
-        if par_u is not None:
-            par_u.scatter_forward_insert()
 
         ############################
         # FCT STEP ON WATER HEIGHT #
         ############################
         logEvent("   FCT Step/Convex Limiting", level=1)
         self.F.FCTStep()
+        if par_u is not None:
+            par_u.scatter_forward_insert()
 
         #############################################
         # UPDATE SOLUTION THROUGH calculateResidual #
         #############################################
         self.F.secondCallCalculateResidual = 1
         self.computeResidual(u,r,b)
-        self.F.check_positivity_water_height=True
+        if par_u is not None:
+            par_u.scatter_forward_insert()
```

Comment:
Could you remind me what `scatter_forward_insert` does?

Reply:
You can take a look at the PetscVec GhostUpdate functionality here: https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Vec/VecGhostUpdateBegin.html

Basially, a parallel vector has a local representation u and a distributed/global representation U. The local representation will have both owned entries and ghost entries owned by other MPI processes.  A forward scatter updates the ghost entries with values from the owned entries and the INSERT mode says to just insert the values from the owned elements. We do this when a process has correctly computed the owned entries of u but not necessarily the ghost entries. So after a scatter_foward_insert, the local representations of U are all correct on all processors.  On the other hand for the residual r (R) with a non-overlapping,  element-based partitioning you usually do a reverse scatter, sending ghost values to the owning process and you do it in ADD mode because you're trying to accumulate all the element contributions to the owning node. After that all the owned entries are correct but not necessarily the ghost entries.